ext/drpttracer/drpttracer.c

```diff
@@ -0,0 +1,579 @@
+}
+
+/***************************************************************************
+ * PMU CONFIG PARSING FUNCTIONS
+ */
+
+#define PT_PMU_PERF_TYPE_FILE "/sys/devices/intel_pt/type"
+#define PT_PMU_EVENTS_CONFIG_DIR "/sys/devices/intel_pt/format"
+
+/* This function is used to get the Intel PT's PMU type. */
+static bool
+parse_pt_pmu_type(OUT uint32_t *pmu_type)
+{
+    /* Read type str from /sys/devices/intel_pt/type. */
+    char *buf = NULL;
+    uint64_t buf_size = 0;
+    if (!read_file_to_buf(PT_PMU_PERF_TYPE_FILE, &buf, &buf_size)) {
+        ERRMSG("failed to read file " PT_PMU_PERF_TYPE_FILE " to buffer\n");
+        return false;
+    }
+
+    /* Parse type str to get the PMU type. */
+    uint32_t type = 0;
+    int ret = dr_sscanf(buf, "%" PRIu32, &type);
+    if (ret < 1) {
+        ERRMSG("failed to parse pmu type from file " PT_PMU_PERF_TYPE_FILE "\n");
+        dr_global_free(buf, buf_size);
+        return false;
+    }
+    *pmu_type = type;
+
+    dr_global_free(buf, buf_size);
+    return true;
+}
+
+/* This function is used to set the PMU event's config.
+ * It first gets the bit field of the 'name' in perf_event_attr.config and then
+ * sets the 'val' to the corresponding bit.
+ */
+static bool
+parse_pt_pmu_event_config(IN const char *name, uint64_t val, OUT uint64_t *config)
+{
+    /* Read config str from /sys/devices/intel_pt/format/name. */
+    char *buf = NULL;
+    uint64_t buf_size = 0;
+    char file_path[MAXIMUM_PATH];
+    dr_snprintf(file_path, MAXIMUM_PATH, PT_PMU_EVENTS_CONFIG_DIR "/%s", name);
+    if (!read_file_to_buf(file_path, &buf, &buf_size)) {
+        ERRMSG("failed to read file %s to buffer\n", file_path);
+        return false;
+    }
+
+    /* Parse config str to get the bit field of the 'name'. */
+    int start = 0, end = 0;
+    int ret = dr_sscanf(buf, "config:%d-%d", &start, &end);
+    if (ret < 1) {
+        ERRMSG("failed to parse event's config bit fields from the format file %s\n",
+               file_path);
+        dr_global_free(buf, buf_size);
+        return false;
+    }
+    if (ret == 1) {
+        end = start;
+    }
+
+#define BITS(x) ((x) == 64 ? -1ULL : (1ULL << (x)) - 1)
+    /* Set the 'val' to the corresponding bit. */
+    uint64_t mask = BITS(end - start + 1);
+    if ((val & ~mask) > 0) {
+        ERRMSG("pmu event's config value %" PRIu64 " is out of range\n", val);
+        dr_global_free(buf, buf_size);
+        return false;
+    }
+    *config |= (val & mask) << start;
+
+    dr_global_free(buf, buf_size);
+    return true;
+}
+
+/***************************************************************************
+ * PERF EVENT RELATED FUNCTIONS
+ */
+
+/* This function is used to initialize one perf_event_attr.
+ * For kernel only tracing:
+ *  The perf_event_attr is set to be the same as running
+ *  'perf record -e intel_pt/noretcomp/k'
+ * For user only tracing:
+ *  The perf_event_attr is set to be the same as running
+ *  'perf record -e intel_pt/cyc,noretcomp/u'
```

Comment:
Explain why we need cyc for user tracing.

Reply:
Done.